### ISSUE TITLE
feat: DeepOmit

### ISSUE THREAD

**Issue by yxliang01** [krzkaczor/ts-essentials#47]
A recursive version of `Omit`. Such type would be useful.

**Comment by krzkaczor** [comment]
@all-contributors please add @yxliang01 for code, idea, docs

**Comment by ThomasdenH** [comment]
What would it do?

**Comment by krzkaczor** [comment]
> A recursive version of `Omit`. Such type would be useful.

How would you specify nested keys? I am not sure if this is possible currently.

**Comment by quezak** [comment]
> How would you specify nested keys? I am not sure if this is possible currently.

![76e4ace506105b098aa24865e9e4a0d4451299a475f5dde880f48ae49a848019](https://user-images.githubusercontent.com/666206/60263244-73898780-98e0-11e9-82fe-aefd566ac44a.jpg)

After some tinkering, I've got this: you specify nested keys via a nested object:
 * keys to be fully omitted are set to `true`
 * keys to be "partially omitted" are specified in a nested object
 * keys to be left untouched are not specified

```typescript
type DeepOmitSpec<T> = {
  [K in keyof T]?: T[K] extends object ? (true | DeepOmitSpec<T[K]>) : true
};

type TrueKeys<T> = ({ [K in keyof T]: T[K] extends true ? K : never })[keyof T];

export type DeepOmit<T, Spec extends DeepOmitSpec<T>> = {
  [K in Exclude<keyof T, TrueKeys<Spec>>]:
    Spec[K] extends DeepOmitSpec<T[K]>
      ? DeepOmit<T[K], Spec[K]>
      : T[K]
};

interface SimpleType1 {
  a: number;
  b: string;
}

interface ComplexType1 {
  c: boolean;
  d: SimpleType1;
  e: SimpleType1;
}

interface ComplexType2 {
  f: string;
  g: SimpleType1;
  h: ComplexType1;
}

type Test0 = DeepOmit<SimpleType1, {}>;
type Test0OK = Assert<IsExact<Test0, SimpleType1>>;

type Test1 = DeepOmit<SimpleType1, { a: true }>;
type Test1OK = Assert<IsExact<Test1, { b: string }>>;

type Test2 = DeepOmit<ComplexType1, { e: true; d: { a: true } }>;
type Test2OK = Assert<IsExact<Test2, { c: boolean; d: { b: string }}>>;

type Test3 = DeepOmit<ComplexType2, {
  f: true,
  g: { a: true };
  h: {
    e: true;
    d: { b: true };
  };
}>
type Test3OK = Assert<IsExact<Test3, {
  g: { b: string };
  h: {
    c: boolean;
    d: { a: number }
  }
}>>;
```

The questions:
 * what are the use cases for this type? are they common enough to add it?
 * is my proposed way to specify omitted keys clean enough to add such a type?

If both are true, I can prepare a PR for that, but I'm unsure about both answers myself. Writing this was a nice brain exercise though :)

**Comment by quezak** [comment]
If the use cases are not that common, you probably can (partially) solve the problem with a combination of `Omit` & `Merge` -- this _should_ produce the same results, though you could accidentally totally change the type of the "partially omitted" key.

```typescript
type Test2 = Merge<Omit<ComplexType1, 'e'>, {
  d: Omit<SimpleType1, 'a'>;
}>
```

**Comment by krzkaczor** [comment]
I really like this implementation! I thought about omitting single keys but your implementation totally makes sense.

> The questions:
> what are the use cases for this type? are they common enough to add it?
> is my proposed way to specify omitted keys clean enough to add such a type?

1) I think that one can always find some use case for such types, probably @yxliang01 can put some more light about initial implementation. 
2) It's awesome 🎉 

I would love to see a PR adding this.

**Comment by lucifer1004** [comment]
I have not read the discussion here, but @quezak, we just came up with the same solution!

**Comment by krzkaczor** [comment]
Closed by #68 avaiable as `ts-essentials@3.0.0-beta.1` (next branch)
